core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,443 @@
+	maxMinTimestampFromNow = int64(300) // 5 minutes
+)
+
+var (
+	bundleGauge = metrics.NewRegisteredGauge("bundlepool/bundles", nil)
+	slotsGauge  = metrics.NewRegisteredGauge("bundlepool/slots", nil)
+)
+
+var (
+	// ErrSimulatorMissing is returned if the bundle simulator is missing.
+	ErrSimulatorMissing = errors.New("bundle simulator is missing")
+
+	// ErrBundleTimestampTooHigh is returned if the bundle's MinTimestamp is too high.
+	ErrBundleTimestampTooHigh = errors.New("bundle MinTimestamp is too high")
+
+	// ErrBundleGasPriceLow is returned if the bundle gas price is too low.
+	ErrBundleGasPriceLow = errors.New("bundle gas price is too low")
+
+	// ErrBundleAlreadyExist is returned if the bundle is already contained
+	// within the pool.
+	ErrBundleAlreadyExist = errors.New("bundle already exist")
+)
+
+// BlockChain defines the minimal set of methods needed to back a tx pool with
+// a chain. Exists to allow mocking the live chain out of tests.
+type BlockChain interface {
+	// Config retrieves the chain's fork configuration.
+	Config() *params.ChainConfig
+
+	// CurrentBlock returns the current head of the chain.
+	CurrentBlock() *types.Header
+
+	// GetBlock retrieves a specific block, used during pool resets.
+	GetBlock(hash common.Hash, number uint64) *types.Block
+
+	// StateAt returns a state database for a given root hash (generally the head).
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+type BundleSimulator interface {
+	SimulateBundle(bundle *types.Bundle) (*big.Int, error)
+}
+
+type BundlePool struct {
+	config    Config
+	mevConfig miner.MevConfig
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	simulator BundleSimulator
+
+	sequencerClients map[string]*rpc.Client
+	deliverBundleCh  chan *types.SendBundleArgs
+	exitCh           chan struct{}
+}
+
+func New(config Config, mevConfig miner.MevConfig) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:          config,
+		mevConfig:       mevConfig,
+		bundles:         make(map[common.Hash]*types.Bundle),
+		bundleHeap:      make(BundleHeap, 0),
+		deliverBundleCh: make(chan *types.SendBundleArgs),
+		exitCh:          make(chan struct{}),
+	}
+	if !pool.mevConfig.BuilderEnabled {
+		return pool
+	}
+	for _, v := range mevConfig.Sequencers {
+		pool.register(v.URL)
+	}
+	if len(pool.sequencerClients) == 0 {
+		log.Error("No valid sequencers")
+	}
+
+	return pool
+}
+
+func (p *BundlePool) deliverLoop() {
+	for {
+		select {
+		case bundle := <-p.deliverBundleCh:
+			for url, cli := range p.sequencerClients {
+				cli := cli
+				url := url
+				go func() {
+					var hash common.Hash
+					err := cli.CallContext(context.Background(), &hash, "eth_sendBundle", *bundle)
+					if err != nil {
+						log.Error("failed to deliver bundle to sequencer", "url", url, "err", err)
```

Comment:
Maybe it's better to add some metrics. We can know in time that a bundle failed to be passed to the sequencer, what is the failure rate, etc.

Reply:
Added, please help review again.